CMakeLists.txt

```diff
@@ -25,15 +25,108 @@ set(PROJECT_VERSION_PATCH 0)
 set(BUILD_SHARED_LIBS ON)
 ## set our cmake module path
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+include(CMakeDependentOption)
+include(CheckIncludeFile)
+include(ExternalProject)
+include(DownloadProject)
+include(CheckCXXCompilerFlag)
+
+
+option(BUILD_LOCAL_DEPENDENCIES "Downloads and builds all dependencies locally " OFF)
+option(HBASE_TARGET_TAG "Downloads and builds all dependencies locally" "e5345b3a7c32c6a80394319c17540b84c8fe66ba")
+option(BUILD_HBASE  "Builds Hbase " OFF)
+
+
+
+######### Includes
+
+	
+if (WIN32)
+	set(BYPRODUCT_SUFFIX ".lib" CACHE STRING "" FORCE)
+	set(BYPRODUCT_SHARED_SUFFIX ".lib" CACHE STRING "" FORCE)
+	set(BYPRODUCT_PREFIX "" CACHE STRING "" FORCE)
+	set(BUILD_ARGS " -GVisual Studio 15 2017")
+else()
+	set(BYPRODUCT_PREFIX "lib" CACHE STRING "" FORCE)
+	set(BYPRODUCT_SHARED_SUFFIX ".so" CACHE STRING "" FORCE)
+	set(BYPRODUCT_SUFFIX ".a" CACHE STRING "" FORCE)
+endif()
+	
+
+
 ## include the Protobuf generation code
 include(ProtobufGen)
+include(DownloadFolly)
+include(DownloadWangle)
+include(DownloadZookeeper)
+
+if (BUILD_LOCAL_DEPENDENCIES)
+	## we want to find the system protoc
+	download_project(PROJ Protobuf
+				 IS_AUTOGEN
+				 GIT_REPOSITORY "https://github.com/protocolbuffers/protobuf.git"
+	  			 GIT_TAG "3.5.1.1")
+	  			 
+	set(PROTOBUF_DIR "${Protobuf_BINARY_DIR}" CACHE STRING "" FORCE)
+	
+	add_library(Protobuf STATIC IMPORTED)
+	set_target_properties(Protobuf PROPERTIES IMPORTED_LOCATION "${Protobuf_BINARY_DIR}/lib/libprotobuf.a" )
+	set(PROTOBUF_LIBS "${Protobuf_BINARY_DIR}/lib/libprotobuf.a" "${Protobuf_BINARY_DIR}/lib/libprotoc.a" CACHE STRING "" FORCE)
+	set(PROTOBUF_INCLUDE_DIRS "${Protobuf_BINARY_DIR}/include" CACHE STRING "" FORCE)
+	add_dependencies(Protobuf Protobuf-download)
+	set(PROTOBUF_FOUND TRUE CACHE STRING "" FORCE)
+	
+	set(PROTOBUF_PROTOC_EXECUTABLE "${Protobuf_BINARY_DIR}/bin/protoc" CACHE STRING "" FORCE)
+	## Add CMAKE_MODULE_PATHS
+	
+	list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/zookeeper/local")
+	list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/protobuf/local")
+	list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/folly/local")
+	list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/wangle/local")
+
+	## Build Apache HBase components that are necessary for this project
+
+	if( BUILD_HBASE )
```

Comment:
IMO, this shouldn't be nested inside BUILD_LOCAL_DEPENDENCIES. General usecase is to keep the regular dependencies static (folly..etc) but change versions of HBase we are building with (for example for newer releases). Thoughts?